### PR TITLE
*: support collect and async load stats for mv index | tidb-test=pr/2210

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref #46539

### What is changed and how it works?

1. For analyzing mv index, use an independent v2 analyze index task to handle it.
    1. When building analyze tasks, add a new `AnalyzeIndexTask` for each mv index. (`planner/core/planbuilder.go`)
    2. When storing the analyze result, specially handle mv index, like the row count, etc. See the comments for details. (`executor/analyze_idx.go`, `statistics/handle/handle.go`, `statistics/histogram.go`, `statistics/analyze.go`)
3. For loading stats, in `initStats()`, we consider all columns instead of pruned columns to make stats for the virtual columns (and therefore expression index, including mv index) able to be async loaded. (`planner/core/stats.go`)
4. For estimation on mv index, we keep it using pseudo estimation to make its behavior unchanged. We'll enable the estimation after it's supported. (`planner/cardinality/row_count_index.go`)

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/handle.go [hunk 1]
```diff
@@ -1125,18 +1125,46 @@ func SaveTableStatsToStorage(sctx sessionctx.Context, results *statistics.Analyz
 	if len(rows) > 0 {
 		snapshot := rows[0].GetUint64(0)
 		// A newer version analyze result has been written, so skip this writing.
```
Comment (winoros) [inline statistics/handle/handle.go:1127]: Update the comment too.

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -2439,14 +2439,17 @@ func getColOffsetForAnalyze(colsInfo []*model.ColumnInfo, colID int64) int {
 // in tblInfo.Indices, index.Columns[i].Offset is set according to tblInfo.Columns. Since we decode row samples according to colsInfo rather than tbl.Columns
 // in the execution phase of ANALYZE, we need to modify index.Columns[i].Offset according to colInfos.
 // TODO: find a better way to find indexed columns in ANALYZE rather than use IndexColumn.Offset
-func getModifiedIndexesInfoForAnalyze(sctx sessionctx.Context, tblInfo *model.TableInfo, allColumns bool, colsInfo []*model.ColumnInfo) []*model.IndexInfo {
+// For multi-valued index, we need to collect it separately here and analyze it as independent index analyze task.
+// See comments for AnalyzeResults.ForMVIndex for more details.
+func getModifiedIndexesInfoForAnalyze(tblInfo *model.TableInfo, allColumns bool, colsInfo []*model.ColumnInfo) ([]*model.IndexInfo, []*model.IndexInfo) {
 	idxsInfo := make([]*model.IndexInfo, 0, len(tblInfo.Indices))
+	independantIdxsInfo := make([]*model.IndexInfo, 0)
```
Comment (winoros) [inline planner/core/planbuilder.go:2446]: typo, ->independent

File: planner/cardinality/row_count_index.go [hunk 3]
```diff
@@ -49,16 +49,20 @@ func GetRowCountByIndexRanges(sctx sessionctx.Context, coll *statistics.HistColl
...
 			name = idx.Info.Name.O
 			for _, col := range idx.Info.Columns {
 				colNames = append(colNames, col.Name.O)
 			}
+			isMVIndex = idx.Info.MVIndex
 		}
 	}
 	recordUsedItemStatsStatus(sctx, idx, coll.PhysicalID, idxID)
-	if !ok || idx.IsInvalid(sctx, coll.Pseudo) {
+	// For the mv index case, now we have supported collecting stats and async loading stats, but sync loading and
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/cardinality/row_count_index.go:63]: The mv index is essentially an expression index, which is an index on virtual columns, which are new columns that come from expression eval results on other columns.
The current sync load mechanism looks for which stats to load by looking into the predicates and seeing what columns they contain. Currently, it doesn't implement things like "see if some expressions behind virtual columns match any predicates" or "see if this predicate contains some columns that are dependent by some virtual columns".

File: statistics/handle/handle.go (hunk 1 above)
Comment (time-and-fate) [reply statistics/handle/handle.go:1127]: Updated.

File: statistics/handle/handle.go [hunk 4]
```diff
@@ -1125,18 +1125,46 @@ func SaveTableStatsToStorage(sctx sessionctx.Context, results *statistics.Analyz
...
+	} else if results.ForMVIndex {
+		// 1-2. There's already an existing record for this table, and we are handling stats for mv index now.
+		// In this case, we only update the version. See comments for AnalyzeResults.ForMVIndex for more details.
+		if _, err = exec.ExecuteInternal(ctx,
+			"update mysql.stats_meta set version=%? where table_id=%?",
+			version,
+			tableID,
+		); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply statistics/handle/handle.go:1167]: Do you mean to update the row count or snapshot?
I explain some reasons in the comments for `AnalyzeResults.ForMVIndex`.

**Review by hawkingrei — APPROVED** [review]

File: planner/cardinality/row_count_index.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (winoros) [reply planner/cardinality/row_count_index.go:63]: I got the problem. Let's fix it later.

**Review by winoros — APPROVED** [review]

**Pull request merged**